src/ir/match.h

```diff
@@ -0,0 +1,782 @@
+template<class T> struct AnyKind {};
+template<class T> struct KindTypeRegistry<AnyKind<T>> {
+  using matched_t = T;
+  using data_t = unused_t;
+};
+template<class T> decltype(auto) Any(T* binder) {
+  return Matcher<AnyKind<T>>(binder, {});
+}
+
+// Exact<T>: matches exact values of the expected type
+template<class T> struct ExactKind {};
+template<class T> struct KindTypeRegistry<ExactKind<T>> {
+  using matched_t = T;
+  using data_t = T;
+};
+template<class T> struct MatchSelf<ExactKind<T>> {
+  bool operator()(T& self, T& expected) { return self == expected; }
+};
+template<class T> decltype(auto) Exact(T* binder, T data) {
+  return Matcher<ExactKind<T>>(binder, data);
+}
+
+// {I32,I64,Int,F32,F64,Float,Number}Lit: match `Literal` of the expected `Type`
+struct I32LK {
+  static bool matchType(Literal& lit) { return lit.type == Type::i32; }
+  static int32_t getVal(Literal& lit) { return lit.geti32(); }
+};
+struct I64LK {
+  static bool matchType(Literal& lit) { return lit.type == Type::i64; }
+  static int64_t getVal(Literal& lit) { return lit.geti64(); }
+};
+struct IntLK {
+  static bool matchType(Literal& lit) { return lit.type.isInteger(); }
+  static int64_t getVal(Literal& lit) { return lit.getInteger(); }
+};
+struct F32LK {
+  static bool matchType(Literal& lit) { return lit.type == Type::f32; }
+  static float getVal(Literal& lit) { return lit.getf32(); }
+};
+struct F64LK {
+  static bool matchType(Literal& lit) { return lit.type == Type::f64; }
+  static double getVal(Literal& lit) { return lit.getf64(); }
+};
+struct FloatLK {
+  static bool matchType(Literal& lit) { return lit.type.isFloat(); }
+  static double getVal(Literal& lit) { return lit.getFloat(); }
+};
+template<class T> struct LitKind {};
+template<class T> struct KindTypeRegistry<LitKind<T>> {
+  using matched_t = Literal;
+  using data_t = unused_t;
+};
+template<class T> struct MatchSelf<LitKind<T>> {
+  bool operator()(Literal& lit, unused_t) { return T::matchType(lit); }
+};
+template<class T> struct NumComponents<LitKind<T>> {
+  static constexpr size_t value = 1;
+};
+template<class T> struct GetComponent<LitKind<T>, 0> {
+  decltype(auto) operator()(Literal lit) { return T::getVal(lit); }
+};
+template<class S> decltype(auto) I32Lit(Literal* binder, S s) {
+  return Matcher<LitKind<I32LK>, S>(binder, {}, s);
+}
+template<class S> decltype(auto) I64Lit(Literal* binder, S s) {
+  return Matcher<LitKind<I64LK>, S>(binder, {}, s);
+}
+template<class S> decltype(auto) IntLit(Literal* binder, S s) {
+  return Matcher<LitKind<IntLK>, S>(binder, {}, s);
+}
+template<class S> decltype(auto) F32Lit(Literal* binder, S s) {
+  return Matcher<LitKind<F32LK>, S>(binder, {}, s);
+}
+template<class S> decltype(auto) F64Lit(Literal* binder, S s) {
+  return Matcher<LitKind<F64LK>, S>(binder, {}, s);
+}
+template<class S> decltype(auto) FloatLit(Literal* binder, S s) {
+  return Matcher<LitKind<FloatLK>, S>(binder, {}, s);
+}
+struct NumberLitKind {};
```

Comment:
> Does this only support i32? Or we only have usecases for i32 and this can be extended to support other types later?

This is able to match i32, i64, f32, or f64, but its data is always an int32_t because that makes it possible to match both ints and floats. The user-facing wrapper function for this matcher is `constant`, and if the user wants to match a more restricted set of types, they can use `i32`, `f32`, `ival`, etc.

> Is that because we want to bypass the type checking?

Yes, exactly. Most of the literal matchers are able to extract a value of a known type, which can then be matched with `Any` or `Exact`. But `NumberLitKind` doesn't know what kind of value it would have to extract and there's no way to additional checking in `GetComponent`, so there's no way to use a submatcher and make it type check. Since we don't have a statically known type for a subcomponent, `NumberLitKind` has to do all its work in `MatchSelf`.